File: scripts/godot-api-drift.mjs
// A fresh dump of the pinned engine, against the vocabulary this repository committed.
//
// Everything else in the suite holds Gofer to Gofer: the catalogue to the addon, the addon to its
// read-backs, a command to its handler. Those are ours, and they change when we change them.
// `protocol/godot-vocabulary.json` is not ours — it is the engine's key names, its performance
// monitors and its Variant types, and every one of them reaches the model as a schema `enum`.
//
// So this is its own gate rather than another test. Nothing in it can fail because of a commit;
// it fails because the pin in `protocol/godot-artifacts.json` moved. The fix for a red run is to
// re-run `npm run generate` against the moved pin and commit what it writes.

import {readFile} from 'node:fs/promises'
import {fileURLToPath} from 'node:url'
import {VOCABULARY_PATH, buildVocabulary, serializeVocabulary} from './godot-vocabulary.mjs'
import {pinnedVersionPrefix, resolveGodotBinary} from './godot-binary.mjs'

const root = fileURLToPath(new URL('..', import.meta.url))
const binary = resolveGodotBinary()

const fresh = await buildVocabulary(binary)
process.stdout.write(
    `Checking ${VOCABULARY_PATH} against Godot ${fresh.engine} (pinned ${pinnedVersionPrefix()})\n`
)

const committed = await readFile(new URL(VOCABULARY_PATH, `file://${root}`), 'utf8')
const written = serializeVocabulary(fresh)
if (committed === written) {
    process.stdout.write(
        `${fresh.keys.length} key names, ${fresh.monitors.length} monitors and `
            + `${fresh.variantTypes.length} Variant types, all as committed\n`
    )
    process.exit(0)
}

const before = JSON.parse(committed)
const named = one => (typeof one === 'string' ? one : one.name)
for (const list of ['keys', 'monitors', 'variantTypes', 'mouseButtons', 'joyButtons', 'joyAxes']) {
    const had = before[list].map(named)
    const has = fresh[list].map(named)
    const gone = had.filter(name => !has.includes(name))
    const added = has.filter(name => !had.includes(name))
    if (gone.length > 0) process.stdout.write(`${list}: gone — ${gone.join(', ')}\n`)
    if (added.length > 0) process.stdout.write(`${list}: new — ${added.join(', ')}\n`)
}
if (before.engine !== fresh.engine)
    process.stdout.write(`engine: ${before.engine} → ${fresh.engine}\n`)

process.stderr.write(
    `${VOCABULARY_PATH} is not what this engine answers. Run \`npm run generate\` and commit what `
        + 'it writes.\n'
)
process.exit(1)

File: src-tauri/src/godot_api_drift.rs
//! What the tool catalog claims about the engine, checked against the engine.
//!
//! Everything else in the suite holds Gofer to Gofer: the catalog to the addon, the addon to its
//! read-backs, a command to its handler. Those are ours, and they change when we change them. The
//! claims here are not ours. `key` names are resolved by `OS.find_keycode_from_string` out of a
//! table that belongs to Godot, and the catalog prints that table to the model as prose — so a
//! sentence that is true today is true until the engine says otherwise, and nothing in this
//! repository would notice the day it does.
//!
//! That is why this is its own gate rather than another acceptance test. Nothing in here can fail
//! because of a commit; it fails because the pinned version in `protocol/godot-artifacts.json`
//! moved. Run it when that pin moves:
//!
//! ```text
//! npm run test:godot:api
//! ```
//!
//! A failure is not a bug to fix in the addon. It is the engine having renamed something the model
//! is being told to send, and the fix is the sentence in `CATALOG`.

use crate::godot_editor_harness::Session;
use crate::tool_params::{GODOT_KEY_NAME, GODOT_KEY_NAME_REFUSED};
use serde_json::json;

/// Every key name the catalog advertises, against the engine that has to know it.
///
/// The catalog is the only place the model is told how to name a key, and it named one the way a
/// browser does — `Return`, which Godot calls `Enter`. The engine refused it, the model had no way
/// to learn what would be accepted, and the turn was spent guessing. The round-trip matters as much
/// as the acceptance: a name the engine resolves but reports back differently would leave
/// `list_input_actions` disagreeing with the sentence that produced it, which is the same trap one
/// step later.
///
/// Both lists are the vocabulary the signature is generated from, so this cannot drift from what
/// the model reads. They used to be parsed back out of the English of a catalogue summary.
#[test]
fn every_key_name_the_catalog_advertises_is_one_the_engine_knows() {
    let (accepted, refused) = (GODOT_KEY_NAME, GODOT_KEY_NAME_REFUSED);
    let session = Session::start();
    for key in accepted {
        let written = session.call(
            "project.set_input_action",
            json!({"name": "drift_keys", "events": [{"kind": "key", "key": key}]}),
        );
        assert_eq!(
            written["events"],
            json!([{"kind": "key", "key": key}]),
            "the catalog offers `{key}`, and this engine answers with something else"
        );
    }
    for key in refused {
        let failure = session.error(
            "project.set_input_action",
            json!({"name": "drift_keys", "events": [{"kind": "key", "key": key}]}),
            None,
        );
        assert!(
            failure.starts_with("unsupported_value"),
            "the catalog says `{key}` is refused, and this engine took it: {failure}"
        );
    }
}
